src/ty.md

```diff
@@ -17,110 +179,378 @@ tcx: TyCtxt<'tcx>
+modules choose to import a larger or smaller set of names explicitly.
 
-You can basically ignore the `TyS` struct – you will basically never
-access it explicitly. We always pass it by reference using the
-`Ty<'tcx>` alias – the only exception I think is to define inherent
-methods on types. Instances of `TyS` are only ever allocated in one of
-the rustc arenas (never e.g. on the stack).
+## ADTs Representation
 
-One common operation on types is to **match** and see what kinds of
-types they are. This is done by doing `match ty.sty`, sort of like this:
+Let’s consider the following example and look at how it is represented:
 
 ```rust,ignore
-fn test_type<'tcx>(ty: Ty<'tcx>) {
-    match ty.sty {
-        ty::TyArray(elem_ty, len) => { ... }
-        ...
-    }
+struct MyStruct<T> { x: u32, y: T }
+```
+
+Here is the definition of `TyKind::Adt`:
+
+```rust,ignore
+Adt(&'tcx AdtDef, SubstsRef<'tcx>)
+```
+
+There are two parts:
+
+- The [`AdtDef`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.AdtDef.html)
+  defines the struct/enum/union without the type parameters. In our example, this is the `MyStruct`
+  part *without* the generic argument `T`.
+    - Note that in the HIR, structs, enums and unions are represented differently, but in `ty::Ty`,
+      they are all represented using `TyKind::Adt`.
+- The
+  [`SubstsRef`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/subst/type.SubstsRef.html)
+  is an interned list of types that are to be substituted for the generic type parameters. For
+  example, if we had a `MyStruct<u32>`, we would end up with a list like `[u32]`. We’ll dig more
+  into generics and substitutions in a little bit.
+
+**`AdtDef` and `DefId`**
+
+For every type defined in the source code, there is a unique `DefId`. This includes ADTs and
+generics. In the example from above, there are two `DefId`s: one for `MyStruct` and one for `T`.
+Notice that the code above does not generate a new `DefId` for `u32` because it is not defined in
+that code (it is only referenced).
+
+`AdtDef` is more or less a wrapper around `DefId` with lots of useful helper methods. There is
+essentially a one-to-one relationship between `AdtDef` and `DefId`. You can get the `AdtDef` for a
+`DefId` with the [`tcx.adt_def(def_id)`
+query](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html#method.adt_def).
+The `AdtDef`s are all interned (as you can see `'tcx` lifetime on it).
+
+We also have an internal map to go from `def_id` to what’s called "Def path". "Def path" is like a
+module path but a bit more rich. For our example, it may be `crate::foo::MyStruct` that identifies
+this definition uniquely. It’s a bit different than a module path because it might include the type
+parameter `T`, which you can't write in normal rust, like `crate::foo::MyStruct::T`. These are used
+in incremental compilation.
+
+### Generics and substitutions
```

Comment:
I wonder if it makes sense to break up this single file into multiple subsections. I tend to think yes (can be a follow-up PR, obviously)

Reply:
:+1: